python/ray/util/client/server/dataservicer.py

```diff
@@ -115,23 +164,57 @@ def Datapath(self, request_iterator, context):
                             req.prep_runtime_env)
                         resp = ray_client_pb2.DataResponse(
                             prep_runtime_env=resp_prep)
+                elif req_type == "connection_cleanup":
+                    cleanup_requested = True
+                    cleanup_resp = ray_client_pb2.ConnectionCleanupResponse()
+                    resp = ray_client_pb2.DataResponse(
+                        connection_cleanup=cleanup_resp)
+                elif req_type == "acknowledge":
+                    # Clean up acknowledged cache entries
+                    response_cache.cleanup(req.acknowledge.req_id)
+                    continue
                 else:
                     raise Exception(f"Unreachable code: Request type "
                                     f"{req_type} not handled in Datapath")
                 resp.req_id = req.req_id
+
+                if should_cache:
+                    response_cache.update_cache(req.req_id, resp)
                 yield resp
-        except grpc.RpcError as e:
-            logger.debug(f"Closing data channel: {e}")
         finally:
             logger.debug(f"Lost data connection from client {client_id}")
-            self.basic_service.release_all(client_id)
             queue_filler_thread.join(QUEUE_JOIN_SECONDS)
             if queue_filler_thread.is_alive():
                 logger.error(
                     "Queue filler thread failed to  join before timeout: {}".
                     format(QUEUE_JOIN_SECONDS))
+            cleanup_delay = self.reconnect_grace_periods.get(client_id)
+            if not cleanup_requested and cleanup_delay is not None:
+                logger.debug("Cleanup wasn't requested, delaying cleanup by"
+                             f"{cleanup_delay} seconds.")
+                # Delay cleanup, since client may attempt a reconnect
+                # Wait on the "stopped" event in case the grpc server is
+                # stopped and we can clean up earlier.
+                self.stopped.wait(timeout=cleanup_delay)
+            else:
+                logger.debug("Cleanup was requested, cleaning up immediately.")
             with self.clients_lock:
-                # Could fail before client accounting happens
+                if client_id not in self.client_last_seen:
+                    logger.debug("Connection already cleaned up.")
+                    # Some other connection has already cleaned up this
+                    # this client's session. This can happen if the client
+                    # reconnects and then gracefully shut's down immediately.
+                    return
+                last_seen = self.client_last_seen[client_id]
+                if last_seen > start_time:
```

Comment:
Can you comment on when self.client_last_seen[client_id] > start_time, and when it is not true?